File: backend/app/schemas/novel.py
# AIMETA P=小说模式_小说和章节请求响应|R=小说结构_章节结构|NR=不含业务逻辑|E=NovelSchema_ChapterSchema|X=internal|A=Pydantic模式|D=pydantic|S=none|RD=./README.ai
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ChoiceOption(BaseModel):
    """前端选择项描述，用于动态 UI 控件。"""

    id: str
    label: str


class UIControl(BaseModel):
    """描述前端应渲染的组件类型与配置。"""

    type: str = Field(..., description="控件类型，如 single_choice/text_input")
    options: Optional[List[ChoiceOption]] = Field(default=None, description="可选项列表")
    placeholder: Optional[str] = Field(default=None, description="输入提示文案")


class ConverseResponse(BaseModel):
    """概念对话接口的统一返回体。"""

    ai_message: str
    ui_control: UIControl
    conversation_state: Dict[str, Any]
    is_complete: bool = False
    ready_for_blueprint: Optional[bool] = None


class ConverseRequest(BaseModel):
    """概念对话接口的请求体。"""

    user_input: Dict[str, Any]
    conversation_state: Dict[str, Any]


class ChapterGenerationStatus(str, Enum):
    NOT_GENERATED = "not_generated"
    GENERATING = "generating"
    EVALUATING = "evaluating"
    SELECTING = "selecting"
    FAILED = "failed"
    EVALUATION_FAILED = "evaluation_failed"
    WAITING_FOR_CONFIRM = "waiting_for_confirm"
    SUCCESSFUL = "successful"


class ChapterOutline(BaseModel):
    chapter_number: int
    title: str
    summary: str
    narrative_phase: Optional[str] = None
    chapter_role: Optional[str] = None
    suspense_hook: Optional[str] = None
    emotional_progression: Optional[str] = None
    character_focus: List[str] = Field(default_factory=list)
    conflict_escalation: List[str] = Field(default_factory=list)
    continuity_notes: List[str] = Field(default_factory=list)
    foreshadowing: Dict[str, List[str]] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ChapterVersionSchema(BaseModel):
    id: Optional[int] = None
    content: str
    style: Optional[str] = "标准"
    evaluation: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: Optional[datetime] = None

class Chapter(ChapterOutline):
    real_summary: Optional[str] = None
    content: Optional[str] = None
    versions: Optional[List[ChapterVersionSchema]] = None
    evaluation: Optional[str] = None
    generation_status: ChapterGenerationStatus = ChapterGenerationStatus.NOT_GENERATED
    word_count: int = 0
    progress_stage: str = "not_generated"
    progress_message: str = ""
    started_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    allowed_actions: List[str] = Field(default_factory=list)
    last_error_summary: Optional[str] = None
    generation_runtime: Optional[Dict[str, Any]] = None


class WorkspaceSummary(BaseModel):
    total_chapters: int = 0
    completed_chapters: int = 0
    failed_chapters: int = 0
    in_progress_chapters: int = 0
    total_word_count: int = 0
    active_chapter: Optional[int] = None
    first_incomplete_chapter: Optional[int] = None
    next_chapter_to_generate: Optional[int] = None
    can_generate_next: bool = False
    available_actions: List[str] = Field(default_factory=list)


class Relationship(BaseModel):
    character_from: str
    character_to: str
    description: str
    core_conflict: Optional[str] = None
    relationship_type: Optional[str] = None
    status: Optional[str] = None
    tension: Optional[str] = None
    direction: Optional[str] = None
    trigger_event: Optional[str] = None
    importance: Optional[int] = None
    extra: Dict[str, Any] = Field(default_factory=dict)


class Blueprint(BaseModel):
    title: str
    target_audience: str = ""
    genre: str = ""
    style: str = ""
    tone: str = ""
    one_sentence_summary: str = ""
    full_synopsis: str = ""
    world_setting: Dict[str, Any] = Field(default_factory=dict)
    characters: List[Dict[str, Any]] = Field(default_factory=list)
    relationships: List[Relationship] = Field(default_factory=list)
    story_arcs: List[Dict[str, Any]] = Field(default_factory=list)
    volume_plan: List[Dict[str, Any]] = Field(default_factory=list)
    foreshadowing_system: List[Dict[str, Any]] = Field(default_factory=list)
    chapter_outline: List[ChapterOutline] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class NovelProject(BaseModel):
    id: str
    user_id: int
    title: str
    initial_prompt: str
    conversation_history: List[Dict[str, Any]] = Field(default_factory=list)
    blueprint: Optional[Blueprint] = None
    chapters: List[Chapter] = Field(default_factory=list)
    generation_runtime: Optional[Dict[str, Any]] = None
    workspace_summary: Optional[WorkspaceSummary] = None

    model_config = ConfigDict(from_attributes=True)


class NovelProjectSummary(BaseModel):
    id: str
    title: str
    genre: str
    last_edited: str
    completed_chapters: int
    total_chapters: int


class BlueprintGenerationResponse(BaseModel):
    blueprint: Blueprint
    ai_message: str


class ChapterGenerationResponse(BaseModel):
    ai_message: str
    chapter_versions: List[Dict[str, Any]]


class NovelSectionType(str, Enum):
    OVERVIEW = "overview"
    WORLD_SETTING = "world_setting"
    CHARACTERS = "characters"
    RELATIONSHIPS = "relationships"
    CHAPTER_OUTLINE = "chapter_outline"
    CHAPTERS = "chapters"


class NovelSectionResponse(BaseModel):
    section: NovelSectionType
    data: Dict[str, Any]


class GenerateChapterRequest(BaseModel):
    chapter_number: int
    writing_notes: Optional[str] = Field(default=None, description="章节额外写作指令")
    quality_requirements: Optional[str] = Field(default=None, description="章节质量偏好与方向")
    target_word_count: Optional[int] = Field(default=None, description="章节目标字数")
    min_word_count: Optional[int] = Field(default=None, description="章节最低字数")


class FlowConfig(BaseModel):
    preset: str = Field(default="basic", description="basic|enhanced|ultimate|longform|custom")
    versions: Optional[int] = Field(default=None, description="生成版本数量")
    enable_preview: Optional[bool] = Field(default=None, description="是否启用预演生成")
    enable_optimizer: Optional[bool] = Field(default=None, description="是否启用优化器")
    enable_consistency: Optional[bool] = Field(default=None, description="是否启用一致性检查")
    enable_enrichment: Optional[bool] = Field(default=None, description="是否启用字数扩写")
    enable_constitution: Optional[bool] = Field(default=None, description="是否启用宪法式写作约束")
    enable_persona: Optional[bool] = Field(default=None, description="是否启用写作人格/风格约束")
    enable_six_dimension: Optional[bool] = Field(default=None, description="是否启用六维评审")
    enable_reader_sim: Optional[bool] = Field(default=None, description="是否启用读者模拟反馈")
    enable_self_critique: Optional[bool] = Field(default=None, description="是否启用自我批判修订")
    enable_memory: Optional[bool] = Field(default=None, description="是否启用记忆层上下文")
    async_finalize: Optional[bool] = Field(default=None, description="是否异步定稿")
    enable_rag: Optional[bool] = Field(default=None, description="是否启用 RAG")
    rag_mode: Optional[str] = Field(default=None, description="simple|two_stage")
    enable_foreshadowing: Optional[bool] = Field(default=None, description="是否启用伏笔约束")
    enable_faction: Optional[bool] = Field(default=None, description="是否启用势力约束")
    target_word_count: Optional[int] = Field(default=None, description="章节目标字数")
    min_word_count: Optional[int] = Field(default=None, description="章节最低字数，低于该值自动扩写")
    max_enrich_iterations: Optional[int] = Field(default=None, description="自动扩写最大迭代次数")
    allow_truncated_response: Optional[bool] = Field(default=None, description="允许模型触发长度截断时返回部分内容")


class AdvancedGenerateRequest(BaseModel):
    project_id: str
    chapter_number: int
    writing_notes: Optional[str] = Field(default=None, description="章节额外写作指令")
    flow_config: FlowConfig = Field(default_factory=FlowConfig)


class AdvancedGenerateVariant(BaseModel):
    index: int
    version_id: int
    content: str
    metadata: Optional[Dict[str, Any]] = None


class AdvancedGenerateResponse(BaseModel):
    project_id: str
    chapter_number: int
    preset: str
    best_version_index: int
    variants: List[AdvancedGenerateVariant]
    review_summaries: Dict[str, Any] = Field(default_factory=dict)
    debug_metadata: Optional[Dict[str, Any]] = None


class FinalizeChapterRequest(BaseModel):
    project_id: str
    selected_version_id: int
    skip_vector_update: Optional[bool] = Field(default=False, description="是否跳过向量库更新")


class FinalizeChapterResponse(BaseModel):
    project_id: str
    chapter_number: int
    selected_version_id: int
    result: Dict[str, Any]


class SelectVersionRequest(BaseModel):
    chapter_number: int
    version_index: int


class EvaluateChapterRequest(BaseModel):
    chapter_number: int
    version_index: Optional[int] = Field(default=None, description="要评审的版本索引（0-based），不传则使用已选版本或最新版本")
    evaluate_all: bool = Field(default=False, description="是否评审所有版本进行对比（多版本评审模式）")


class CancelChapterRequest(BaseModel):
    chapter_number: int
    reason: Optional[str] = Field(default=None, description="手动终止原因")


class UpdateChapterOutlineRequest(BaseModel):
    chapter_number: int
    title: str
    summary: str


class RewriteChapterOutlineRequest(BaseModel):
    chapter_number: int
    title: str
    summary: str
    direction: Optional[str] = Field(default=None, description="重写方向要求")


class DeleteChapterRequest(BaseModel):
    chapter_numbers: List[int]


class GenerateOutlineRequest(BaseModel):
    start_chapter: int
    num_chapters: int
    target_total_chapters: Optional[int] = Field(default=None, description="全书目标总章节数")
    target_total_words: Optional[int] = Field(default=None, description="全书目标总字数")
    chapter_word_target: Optional[int] = Field(default=None, description="单章目标字数")


class BlueprintPatch(BaseModel):
    one_sentence_summary: Optional[str] = None
    full_synopsis: Optional[str] = None
    world_setting: Optional[Dict[str, Any]] = None
    characters: Optional[List[Dict[str, Any]]] = None
    relationships: Optional[List[Relationship]] = None
    story_arcs: Optional[List[Dict[str, Any]]] = None
    volume_plan: Optional[List[Dict[str, Any]]] = None
    foreshadowing_system: Optional[List[Dict[str, Any]]] = None
    chapter_outline: Optional[List[ChapterOutline]] = None


class EditChapterRequest(BaseModel):
    chapter_number: int
    content: str

File: frontend/src/api/types/novel.ts
// AIMETA P=小说类型定义|R=类型导出|NR=不含API逻辑|E=types:novel|X=internal|D=none|S=none|RD=./README.ai
// Phase 5.2 重构：从 novel.ts 提取的所有 TypeScript 类型定义
// 此文件只包含 interface/type/class ApiError，不包含任何 API 调用逻辑

// ============================================================================
// 错误处理类型（ApiError class 定义在 novel-client.ts 中，因为其 constructor 依赖 formatApiErrorMessage）
// ============================================================================

export interface ApiErrorDetail {
  status: number
  message: string
  code?: string
  hint?: string
  rootCause?: string
  requestId?: string
  retryable?: boolean
  responseSnippet?: string
  rejectionSummary?: Record<string, any>
  missingChapters?: number[]
}

// ============================================================================
// 核心小说类型
// ============================================================================

export interface NovelProject {
  id: string
  title: string
  initial_prompt: string
  blueprint?: Blueprint
  chapters: Chapter[]
  conversation_history: ConversationMessage[]
  generation_runtime?: GenerationRuntime
  workspace_summary?: WorkspaceSummary
}

export interface NovelProjectSummary {
  id: string
  title: string
  genre: string
  last_edited: string
  completed_chapters: number
  total_chapters: number
}

export interface WorldSetting {
  core_rules?: string
  era?: string
  time_period?: string
  atmosphere?: string
  tone?: string
  era_background?: unknown
  world_structure?: unknown
  power_system?: unknown
  survival_system?: unknown
  life_system?: unknown
  culture_system?: unknown
  civilization_system?: unknown
  economy_system?: unknown
  social_structure?: unknown
  technology_system?: unknown
  resource_system?: unknown
  belief_system?: unknown
  geography_system?: unknown
  faction_order?: unknown
  system_blueprint?: Record<string, unknown>
  key_locations?: Array<Record<string, unknown>>
  factions?: Array<Record<string, unknown>>
  [key: string]: unknown
}

export interface StoryArc {
  title?: string
  theme?: string
  goal?: string
  conflict?: string
  summary?: string
  [key: string]: unknown
}

export interface VolumePlanItem {
  volume?: number | string
  title?: string
  focus?: string
  goal?: string
  summary?: string
  [key: string]: unknown
}

export interface NovelOutlineStage {
  stage?: number
  title?: string
  core_theme?: string
  goal?: string
  main_conflict?: string
  background?: string
  character_progression?: string
  world_progression?: string
  faction_progression?: string
  power_progression?: string
  survival_and_life_progression?: string
  cultural_and_civilizational_progression?: string
  resource_and_operation_line?: string
  emotional_core?: string
  major_setpiece?: string
  story_function?: string
  key_events?: string[]
  turning_points?: string[]
  stage_tasks?: string[]
  stage_climax?: string
  foreshadowing_and_payoff?: string
  ending_hook?: string
  expected_chapter_range?: string
  [key: string]: unknown
}

export interface BlueprintForeshadowingItem {
  plant?: unknown
  payoff?: unknown
  owner?: string
  trigger?: string
  summary?: string
  [key: string]: unknown
}

export interface Blueprint {
  title?: string
  target_audience?: string
  genre?: string
  style?: string
  tone?: string
  one_sentence_summary?: string
  full_synopsis?: string
  world_setting?: WorldSetting
  characters?: Character[]
  relationships?: Relationship[]
  story_arcs?: StoryArc[]
  volume_plan?: VolumePlanItem[]
  novel_outline?: NovelOutlineStage[]
  foreshadowing_system?: BlueprintForeshadowingItem[]
  chapter_outline?: ChapterOutline[]
}

export interface Character {
  name: string
  description?: string
  summary?: string
  role?: string
  identity?: string
  archetype?: string
  personality?: string
  goals?: string
  core_motivation?: string
  fear_or_wound?: string
  external_goal?: string
  hidden_secret?: string
  growth_arc?: string
  first_highlight_chapter?: number | string
  relationship_hook?: string
  importance?: 'protagonist' | 'core' | 'supporting' | 'minor' | string
  tags?: string[]
  abilities?: string
  relationship_to_protagonist?: string
  extra?: Record<string, any>
}

export interface Relationship {
  character_from: string
  character_to: string
  description: string
  relation_type?: string
  relationship_type?: string
  status?: string
  current_state?: string
  core_conflict?: string
  tension?: string
  direction?: string
  expected_change?: string
  trigger_event?: string
  key_trigger?: string
  importance?: number
  extra?: Record<string, any>
}

export interface ChapterOutline {
  chapter_number: number
  title: string
  summary: string
  narrative_phase?: string
  chapter_role?: string
  suspense_hook?: string
  emotional_progression?: string
  character_focus?: string[]
  conflict_escalation?: string[]
  continuity_notes?: string[]
  foreshadowing?: {
    plant?: string[]
    payoff?: string[]
  }
  metadata?: Record<string, any>
}

export interface ChapterVersion {
  id?: number
  content: string
  style?: string
  evaluation?: string
  metadata?: Record<string, any>
}

export interface Chapter {
  chapter_number: number
  title: string
  summary: string
  content: string | null
  selected_version_id?: number | null
  versions: ChapterVersion[] | null
  evaluation: string | null
  generation_status: 'not_generated' | 'generating' | 'evaluating' | 'selecting' | 'failed' | 'evaluation_failed' | 'waiting_for_confirm' | 'successful'
  word_count?: number
  progress_stage?: 'queued' | 'generating' | 'evaluating' | 'selecting' | 'ready' | 'failed' | string
  progress_message?: string | null
  started_at?: string | null
  updated_at?: string | null
  allowed_actions?: string[]
  last_error_summary?: string | null
  generation_runtime?: GenerationRuntime
}

export interface GenerationRuntimeEvent {
  at?: string
  stage?: string
  level?: 'info' | 'warning' | 'error' | string
  kind?: 'status' | 'content' | 'review' | 'continuity' | 'save' | 'error' | string
  title?: string
  summary?: string
  content_preview?: string
  progress_percent?: number
  metrics?: Record<string, any>
  artifact_refs?: Record<string, any> | Array<Record<string, any>>
  developer_detail?: Record<string, any>
  message?: string
  metadata?: Record<string, any>
  /** 正文分片：只有后端 content_delta 事件才会带，日志事件必须为空。 */
  content_delta?: string
  /** 该分片是否只是预览（预览可被后续分片替换，非预览要按段累积）。 */
  content_is_preview?: boolean
  /** 长篇分段序号，用于按段去重累积。 */
  segment_index?: number
}

export interface GenerationRuntime {
  queued?: boolean
  generation_mode?: string
  preset?: string
  version_count?: number
  target_word_count?: number
  min_word_count?: number
  progress_stage?: string
  progress_message?: string
  progress_percent?: number
  estimated_remaining_seconds?: number
  started_at?: string | null
  updated_at?: string | null
  chapter_number?: number
  allowed_actions?: string[]
  last_error_summary?: string | null
  task_id?: string | null
  task_status?: string | null
  task_stage?: string | null
  task_message?: string | null
  event_cursor?: number
  task_event_cursor?: number
  retry_count?: number
  max_retries?: number
  lease_owner?: string | null
  heartbeat_at?: string | null
  elapsed_ms?: number | null
  input_tokens?: number | null
  output_tokens?: number | null
  total_tokens?: number | null
  recovered_from_reload?: boolean
  events?: GenerationRuntimeEvent[]
  [key: string]: any
}

export interface WorkspaceSummary {
  total_chapters: number
  completed_chapters: number
  failed_chapters: number
  in_progress_chapters: number
  total_word_count: number
  active_chapter?: number | null
  first_incomplete_chapter?: number | null
  next_chapter_to_generate?: number | null
  can_generate_next: boolean
  available_actions: string[]
}

export interface ConversationMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface ConverseResponse {
  ai_message: string
  ui_control: UIControl
  conversation_state: any
  is_complete: boolean
  ready_for_blueprint?: boolean
}

export interface BlueprintGenerationResponse {
  blueprint: Blueprint
  ai_message: string
}

export interface BlueprintGenerationError {
  code: string
  message: string
  detail?: string | null
  retryable?: boolean
}

export interface BlueprintGenerationJobResponse {
  run_id: string
  project_id: string
  status: 'idle' | 'queued' | 'generating' | 'polishing' | 'successful' | 'failed' | 'cancelled'
  progress_stage: string
  progress_message: string
  started_at?: string | null
  updated_at?: string | null
  blueprint?: Blueprint | null
  ai_message?: string | null
  error?: BlueprintGenerationError | string | null
}

export interface OutlineGenerationJobResponse {
  run_id: string
  project_id: string
  status:
    | 'idle'
    | 'queued'
    | 'generating'
    | 'outline_context'
    | 'outline_chapter_skeleton'
    | 'outline_rewrite'
    | 'saving'
    | 'successful'
    | 'failed'
    | 'cancelled'
  progress_stage: string
  progress_message: string
  started_at?: string | null
  updated_at?: string | null
  project?: NovelProject | null
  events?: GenerationRuntimeEvent[]
  error?: BlueprintGenerationError | string | null
}

export interface StyleProfileJobResponse {
  run_id: string
  project_id: string
  status: 'idle' | 'queued' | 'extracting' | 'profiling' | 'saving' | 'successful' | 'failed' | 'cancelled'
  progress_stage: string
  progress_message: string
  started_at?: string | null
  updated_at?: string | null
  profile?: any | null
  error?: BlueprintGenerationError | string | null
}

export interface StyleSourceUploadJobResponse {
  run_id: string
  project_id: string
  status:
    | 'idle'
    | 'queued'
    | 'upload_reading'
    | 'upload_extracting'
    | 'upload_saving'
    | 'successful'
    | 'failed'
    | 'cancelled'
  progress_stage: string
  progress_message: string
  started_at?: string | null
  updated_at?: string | null
  filename?: string | null
  source?: any | null
  metrics?: Record<string, any>
  error?: BlueprintGenerationError | string | null
}

export interface NovelImportJobResponse {
  run_id: string
  status:
    | 'idle'
    | 'queued'
    | 'import_reading'
    | 'import_splitting'
    | 'import_sampling'
    | 'import_character_verify'
    | 'import_blueprint_extract'
    | 'import_saving'
    | 'import_ledger_rebuild'
    | 'successful'
    | 'failed'
    | 'cancelled'
  progress_stage: string
  progress_message: string
  started_at?: string | null
  updated_at?: string | null
  filename?: string | null
  project_id?: string | null
  metrics?: Record<string, any>
  error?: BlueprintGenerationError | string | null
}

export interface UIControl {
  type: 'single_choice' | 'multi_choice' | 'text_input'
  options?: Array<{ id: string; label: string }>
  placeholder?: string
}

export interface ChapterGenerationResponse {
  versions: ChapterVersion[]
  evaluation: string | null
  ai_message: string
  chapter_number: number
  generation_runtime?: GenerationRuntime
}

export interface GenerateOutlineOptions {
  targetTotalChapters?: number
  targetTotalWords?: number
  chapterWordTarget?: number
  /** 长篇分卷模式：卷数 */
  volumeCount?: number
  /** 长篇分卷模式：每卷章节数 */
  chaptersPerVolume?: number
  /** 强制走长篇大纲生成器 */
  longForm?: boolean
}

export interface GenerateChapterOptions {
  writingNotes?: string
  qualityRequirements?: string
  minWordCount?: number
  targetWordCount?: number
  segmentWordLimit?: number
  generationTimeoutSeconds?: number
  preset?: 'basic' | 'enhanced' | 'longform' | 'ultimate'
  enableConsistency?: boolean
  enableEnrichment?: boolean
  enableSelfCritique?: boolean
  enableReaderSim?: boolean
  enableMemory?: boolean
  enableForeshadowing?: boolean
  versionCount?: number
  force?: boolean
}

export interface CancelChapterOptions {
  reason?: string
}

export interface RewriteChapterOutlineOptions {
  direction?: string
}

export interface DeleteNovelsResponse {
  status: string
  message: string
}

// ============================================================================
// Section 类型
// ============================================================================

export type NovelSectionType = 'overview' | 'world_setting' | 'novel_outline' | 'characters' | 'relationships' | 'chapter_outline' | 'chapters'

export type AnalysisSectionType =
  | 'emotion_curve'
  | 'foreshadowing'
  | 'knowledge_graph'
  | 'story_trajectory'
  | 'creative_guidance'
  | 'comprehensive_analysis'
  | 'clue_tracker'

export type FeatureEntryType = 'style_learning' | 'memory_management' | 'token_budget' | 'research' | 'clue_tracker' | 'knowledge_graph' | 'foreshadowing'

export type AllSectionType = NovelSectionType | AnalysisSectionType | FeatureEntryType

export interface NovelSectionResponse {
  section: NovelSectionType
  data: Record<string, any>
}

// ============================================================================
// 优化相关类型
// ============================================================================

export interface EmotionBeat {
  primary_emotion: string
  intensity: number
  curve: {
    start: number
    peak: number
    end: number
  }
  turning_point: string
}

export interface OptimizeRequest {
  project_id: string
  chapter_number: number
  dimension:
    | 'dialogue'
    | 'environment'
    | 'psychology'
    | 'rhythm'
  additional_notes?: string
  version_index?: number
  version_id?: number
}

export interface OptimizeResponse {
  optimized_content: string
  optimization_notes: string | string[]
  dimension: string
}

export interface ApplyOptimizationResponse {
  status: string
  message: string
  chapter: Chapter
}

// ============================================================================
// 分析相关类型
// ============================================================================

export interface EnhancedEmotionPoint {
  chapter_number: number
  chapter_id: string
  title: string
  primary_emotion: string
  primary_intensity: number
  secondary_emotions: Array<[string, number]>
  narrative_phase: string
  pace: string
  is_turning_point: boolean
  turning_point_type?: string | null
  description: string
}

export interface StoryTrajectoryAnalysis {
  project_id: string
  project_title: string
  shape: string
  shape_confidence: number
  total_chapters: number
  avg_intensity: number
  intensity_range: [number, number]
  volatility: number
  peak_chapters: number[]
  valley_chapters: number[]
  turning_points: number[]
  description: string
  recommendations: string[]
}

export interface GuidanceItemAnalysis {
  type: string
  priority: string
  title: string
  description: string
  specific_suggestions: string[]
  affected_chapters: number[]
  examples: string[]
}

export interface CreativeGuidanceAnalysis {
  project_id: string
  project_title: string
  current_chapter: number
  overall_assessment: string
  strengths: string[]
  weaknesses: string[]
  guidance_items: GuidanceItemAnalysis[]
  next_chapter_suggestions: string[]
  long_term_planning: string[]
}

export interface ComprehensiveAnalysis {
  project_id: string
  project_title: string
  emotion_points: EnhancedEmotionPoint[]
  trajectory: StoryTrajectoryAnalysis
  guidance: CreativeGuidanceAnalysis
}

// ============================================================================
// 线索追踪类型
// ============================================================================

export interface ForeshadowingItem {
  id: number
  name?: string | null
  chapter_number: number
  content: string
  type: string
  status: string
  target_reveal_chapter?: number | null
  reveal_method?: string | null
  reveal_impact?: string | null
  related_characters?: string[] | null
  related_plots?: string[] | null
  importance?: string | null
  urgency?: number | null
  keywords?: string[] | null
  resolved_chapter_number: number | null
  is_manual: boolean
  ai_confidence: number | null
  author_note: string | null
  created_at: string
}

export interface ForeshadowingListResponse {
  total: number
  limit: number
  offset: number
  data: ForeshadowingItem[]
}

export interface ForeshadowingCreateRequest {
  chapter_id: number
  chapter_number: number
  content: string
  type: string
  keywords?: string[]
  author_note?: string
}

export interface ForeshadowingCreateResponse extends ForeshadowingItem {
  project_id: string
}

export interface ForeshadowingResolveRequest {
  resolved_chapter_id: number
  resolved_chapter_number: number
  resolution_text: string
  resolution_type?: string
  quality_score?: number
}

export interface ForeshadowingResolveResponse {
  status: string
  message: string
  resolution_id: number
}

export interface ForeshadowingReminderItem {
  id: number
  foreshadowing_id: number
  reminder_type: string
  message: string
  status: string
  suggested_chapter_range?: { start?: number; end?: number } | null
  created_at: string
}


// ===== Research Types =====
export type ResearchMode = 'auto' | 'ask' | 'off'
export type ResearchScope = 'global' | 'enhanced' | 'chapter'

export interface ResearchConfigUpdate {
  mode: ResearchMode
  enabled: boolean
  search_provider: 'tavily' | 'serper' | 'bing' | 'none'
  search_base_url?: string | null
  search_api_key?: string | null
  clear_search_api_key?: boolean
  research_llm_base_url?: string | null
  research_llm_model?: string | null
  research_llm_api_key?: string | null
  clear_research_llm_api_key?: boolean
  reuse_writing_llm: boolean
  local_model_enabled: false
  global_research_enabled: boolean
  enhanced_research_enabled: boolean
  chapter_research_enabled: boolean
  max_parallel_queries: number
  max_results_per_query: number
  preferred_domains: string[]
  blocked_domains: string[]
  category_preferences: string[]
}

export interface ResearchConfig extends ResearchConfigUpdate {
  project_id: string
  search_api_key_masked?: string | null
  search_api_key_configured: boolean
  research_llm_api_key_masked?: string | null
  research_llm_api_key_configured: boolean
  provider_priority: string[]
}

export interface ResearchSource {
  url: string
  title?: string | null
  snippet?: string | null
  credibility_score?: number | null
  trust_tier?: string | null
  cross_source_count?: number | null
  [key: string]: unknown
}

export interface ResearchArtifact {
  id: number
  run_id: string
  project_id: string
  scope: ResearchScope
  chapter_number?: number | null
  status: string
  trigger: string
  query_plan: Array<Record<string, unknown>>
  sources: ResearchSource[]
  category_payload: Record<string, unknown>
  summary?: string | null
  file_manifest: Record<string, unknown>
  provider_metadata: Record<string, unknown>
  error?: Record<string, unknown> | null
  started_at?: string | null
  finished_at?: string | null
}

export interface ResearchRunRequest {
  scope: ResearchScope
  chapter_number?: number
  consent: boolean
  force: boolean
  trigger: string
  context?: Record<string, unknown>
}

export interface ResearchJobRead {
  run_id: string
  project_id: string
  scope: ResearchScope
  chapter_number?: number | null
  status: string
  cancel_signal_sent: boolean
  in_process_task_cancelled: boolean
  artifact?: ResearchArtifact | null
}

/** @deprecated Use ResearchJobRead. */
export type ResearchRunStatus = ResearchJobRead

// ===== Research Archive Types =====
export interface ResearchArchive {
  id: string
  project_id: string
  title: string
  content: string
  excerpt?: string | null
  category?: string | null
  source_url?: string | null
  created_at: string
  updated_at?: string | null
}

export interface ResearchArchiveCreate {
  title: string
  content: string
  category?: string | null
  source_url?: string | null
}
